apps/builder/app/canvas/canvas.tsx

```diff
@@ -102,6 +104,24 @@ const useElementsTree = (components: Components, params: Params) => {
       propsByInstanceIdStore,
       assetsStore,
       pagesStore: pagesMapStore,
+      dataSourceValuesStore: computed(
+        [dataSourcesStore, dataSourceValuesStore],
+        (dataSources, dataSourceValues) => {
+          const newValues = new Map<DataSource["id"], unknown>();
+          for (const [dataSourceId, dataSource] of dataSources) {
+            newValues.set(dataSourceId, dataSource.value);
+          }
+          for (const [dataSourceId, dataSourceValue] of dataSourceValues) {
+            newValues.set(dataSourceId, dataSourceValue);
+          }
+          return newValues;
+        }
+      ),
+      onDataSourceUpdate: (dataSourceId, value) => {
+        const dataSourceValues = new Map(dataSourceValuesStore.get());
```

Comment:
Looks like its only to force recompute? To emulate immutable data structure? Mb use immer instead?

Reply:
No big deal to use immer here